File: IDI Open/solutions/2012/inverse/inverse_rs.c
/*
	solution by ruben spaans
	precalculating, smart backtracking
	slower isprime()
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#if defined(__MINGW32__) || defined(__MINGW64__)
	#define LONGLONG "%I64d"
#else
	#define LONGLONG "%lld"
#endif
#define MAXP 1000020
typedef long long ll;
typedef unsigned long long ull;
typedef unsigned int uint;

char sieve[MAXP];
int prime[80000];
int primes;

#define MAXF 90000
int prod[MAXF],sumdiv[MAXF],prim[MAXF];
int sn;

void createsieve() {
  int i,j,q;
  memset(sieve,1,sizeof(sieve));
  q=sqrt(MAXP);
  for(sieve[0]=sieve[1]=0,i=2;i<q;i++) if(sieve[i])
    for(j=i*i;j<MAXP;j+=i) sieve[j]=0;
}

void genprimes() {
	int i;
	for(primes=i=0;i<MAXP;i++) if(sieve[i]) prime[primes++]=i;
}

#define MAXN 1000020
void genfactors() {
	int i,nevner;
	ll teller;
	for(sn=i=0;i<primes;i++) {
		teller=(ll)prime[i]*prime[i];
		nevner=prime[i]-1;
		while((teller-1)/nevner<=MAXN) {
			prod[sn]=teller/prime[i];
			sumdiv[sn]=(teller-1)/nevner;
			prim[sn++]=prime[i];
			teller*=prime[i];
		}
	}
}

int isprime(ull n) {
	int i;
	for(i=0;i<primes && (ll)prime[i]*prime[i]<n;i++)
		if(n%prime[i]<1) return 0;
	return 1;
}

int list[MAXF],listf[MAXF],listp[MAXF];
int ln;

ll ans[1000000];
int nn;

int compi(const void *A,const void *B) {
	const ll *a=A,*b=B;
	if(*a<*b) return -1;
	if(*a>*b) return 1;
	return 0;
}

void btr(int ix,int at,ll orinum,ll divsum) {
	if(divsum<2) { ans[nn++]=orinum; return; }
	if(divsum-1>=MAXP && divsum-1>at && isprime(divsum-1))
		ans[nn++]=orinum*(divsum-1);
	for(;ix<ln;ix++) if(listp[ix]>at && divsum%listf[ix]<1)
		btr(ix+1,listp[ix],orinum*list[ix],divsum/listf[ix]);
}

int main() {
	int T,i;
	ull n;
	char s[32];
	createsieve();
	genprimes();
	genfactors();
	scanf("%d",&T);
	while(T--) {
		scanf("%s",s);
		for(n=i=0;s[i];i++) n=n*10+s[i]-48;
		for(ln=i=0;i<sn;i++) if(n%sumdiv[i]<1)
			list[ln]=prod[i],listf[ln]=sumdiv[i],listp[ln++]=prim[i];
		nn=0;
		btr(0,0,1,n);
		qsort(ans,nn,sizeof(ll),compi);
		if(!nn) puts("none!");
		else for(i=0;i<nn;i++) printf(LONGLONG"%c",ans[i],i==nn-1?'\n':' ');
	}
	return 0;
}

File: IDI Open/solutions/2012/inverse/inverse_rs.java
import java.io.*;
import java.math.*;
import java.util.*;

public class inverse_rs {
	BufferedReader in=new BufferedReader(new InputStreamReader(System.in));
	StringTokenizer st=new StringTokenizer("");
	String LINE() throws Exception { return in.readLine(); }
	String STR() throws Exception {
		while(!st.hasMoreTokens()) st=new StringTokenizer(LINE());
		return st.nextToken();
	}
	int INT() throws Exception { return Integer.parseInt(STR()); }
	long LONG() throws Exception { return Long.parseLong(STR()); }
	double DOUBLE() throws Exception { return Double.parseDouble(STR()); }
	String PD(double d,int n) {
		return String.format("%."+n+"f",d).replace(",",".");
	}

	public static void main(String args[]) throws Exception {
		new inverse_rs().go();
	}

	boolean[] createsieve(int n) {
		boolean s[] = new boolean[n];
		for(int i=2; i*i<n; i++) if(!s[i]) for(int j=i*i; j<n; j+=i) s[j]=true;
		return s;
	}
	int[] getprimes(boolean s[]) {
		int n=0;
		for(int i=2; i<s.length; i++) if(!s[i]) n++;
		int p[]=new int[n], q=0;
		for(int i=2; i<s.length; i++) if(!s[i]) p[q++]=i;
		return p;
	}
	boolean sieve[];
	int primes[];
	int prod[]=new int[1000000];	/*	prime power */
	int sumdiv[]=new int[1000000];	/*	sum of divisors of this prime power */
	int prim[]=new int[1000000];	/*	base prime */
	int nn;
	ArrayList<Integer> ans;
	
	boolean isprime(int n) {
		if(n<1000000) return sieve[n];
		for(int i=0;primes[i]*primes[i]<n;i++) if(n%primes[i]<1) return false;
		return true;
	}
	void btr(int ix,int at,int orinum,int divsum) {
		if(divsum<2) {
			ans.add(orinum);
			return;
		}
		if(divsum-1>999999 && divsum-1>at && isprime(divsum-1))
			ans.add(orinum*(divsum-1));
		for(;ix<nn;ix++) if(prim[ix]>at && divsum%sumdiv[ix]<1)
			btr(ix+1,prim[ix],orinum*prod[ix],divsum/sumdiv[ix]);
	}
	public void go() throws Exception {
		int T=INT();
		sieve=createsieve(1000000);
		primes=getprimes(sieve);
		while(T-->0) {
			int n=INT();
			nn=0;
			/*	check all eligible divisors for each prime */
			for(int i=0;i<primes.length && primes[i]<=n;i++) {
				long teller=(long)primes[i]*primes[i];
				long nevner=primes[i]-1;
				while((teller-1)/nevner<=n) {
					long z=(teller-1)/nevner;
					if(n%z<1) {
						prod[nn]=(int)(teller/primes[i]);
						sumdiv[nn]=(int)z;
						prim[nn++]=primes[i];
					}
					teller*=primes[i];
				}
			}
			/*	ready to backtrack */
			ans=new ArrayList<Integer>();
			btr(0,0,1,n);
			if(ans.size()==0) System.out.println("none!");
			else {
				Collections.sort(ans);
				StringBuilder b=new StringBuilder(String.format("%d",ans.get(0)));
				int antall=ans.size();
				for(int i=1;i<antall;i++) b.append(String.format(" %d",ans.get(i)));
				System.out.println(b.toString());
			}
		}
	}
}
